--- busybox-1.10.4/sysklogd/syslogd.c.orig	2008-06-25 12:55:25.000000000 +0200
+++ busybox-1.10.4/sysklogd/syslogd.c	2008-07-09 10:14:02.000000000 +0200
@@ -505,8 +505,8 @@
 
 	/* Unlink old /dev/log or object it points to. */
 	/* (if it exists, bind will fail) */
-	strcpy(sunx.sun_path, "/dev/log");
-	dev_log_name = xmalloc_follow_symlinks("/dev/log");
+	strcpy(sunx.sun_path, _PATH_LOG);
+	dev_log_name = xmalloc_follow_symlinks(_PATH_LOG);
 	if (dev_log_name) {
 		safe_strncpy(sunx.sun_path, dev_log_name, sizeof(sunx.sun_path));
 		free(dev_log_name);
@@ -515,7 +515,7 @@
 
 	sock_fd = xsocket(AF_UNIX, SOCK_DGRAM, 0);
 	xbind(sock_fd, (struct sockaddr *) &sunx, sizeof(sunx));
-	chmod("/dev/log", 0666);
+	chmod(_PATH_LOG, 0666);
 
 	return sock_fd;
 }
