--- busybox-1.4.2.orig/networking/wget.c	2007-03-18 19:00:17.000000000 +0100
+++ busybox-1.4.2/networking/wget.c	2007-05-22 09:55:24.000000000 +0200
@@ -92,10 +92,9 @@
 	struct host_info server, target;
 	len_and_sockaddr *lsa;
 	int n, status;
-	int port;
 	int try = 5;
 	unsigned opt;
-	char *s;
+	char *s, *p;
 	char *proxy = 0;
 	char *dir_prefix = NULL;
 #if ENABLE_FEATURE_WGET_LONG_OPTIONS
@@ -423,16 +422,24 @@
 		// Response is "227 garbageN1,N2,N3,N4,P1,P2[)garbage]
 		// Server's IP is N1.N2.N3.N4 (we ignore it)
 		// Server's port for data connection is P1*256+P2
-		s = strrchr(buf, ')');
-		if (s) s[0] = '\0';
-		s = strrchr(buf, ',');
-		if (!s) goto pasv_error;
-		port = xatou_range(s+1, 0, 255);
-		*s = '\0';
-		s = strrchr(buf, ',');
-		if (!s) goto pasv_error;
-		port += xatou_range(s+1, 0, 255) * 256;
-		set_nport(lsa, htons(port));
+		if (!(s = strrchr(buf, ')')))
+			goto pasv_error;
+		*s = 0;	// clear trailing '\0'
+		if (!(s = strrchr(buf, '(')))
+			goto pasv_error;
+		p = ++s;
+		for (n=0; n<4; n++) {
+			if (!(s = strchr(s, ',')))
+				goto pasv_error;
+			*s++ = '.';
+		}
+		s[-1] = 0;
+		if (inet_aton(p, &lsa->sin.sin_addr) == 0)
+			goto pasv_error;
+		if (!(s = strchr((p=s), ',')))
+			goto pasv_error;
+		*s++ = 0;
+		set_nport(lsa, htons(xatou_range(p, 0, 255) * 256 + xatou_range(s, 0, 255)));
 		dfp = open_socket(lsa);
 
 		if (beg_range) {
