--- busybox-1.15.2/init/init.c.orig	2009-10-08 00:59:09.000000000 +0000
+++ busybox-1.15.2/init/init.c	2009-11-14 21:41:41.000000000 +0000
@@ -516,6 +516,9 @@
 	parser_t *parser = config_open2("/etc/inittab", fopen_for_read);
 
 	if (parser == NULL)
+		parser = config_open2("/usr/share/defaults/etc/inittab", fopen_for_read);
+
+	if (parser == NULL)
 #endif
 	{
 		/* No inittab file - set up some default behavior */
