--- busybox/networking/wget.c.orig	2007-11-10 02:40:47.000000000 +0100
+++ busybox/networking/wget.c	2007-11-20 15:43:46.000000000 +0100
@@ -397,11 +397,11 @@
 	char buf[512];
 	struct host_info server, target;
 	len_and_sockaddr *lsa;
-	int status;
+	int nn,status;
 	int port;
 	int try = 5;
 	unsigned opt;
-	char *str;
+	char *str, *p;
 	char *proxy = 0;
 	char *dir_prefix = NULL;
 #if ENABLE_FEATURE_WGET_LONG_OPTIONS
@@ -713,16 +713,24 @@
 		// Response is "227 garbageN1,N2,N3,N4,P1,P2[)garbage]
 		// Server's IP is N1.N2.N3.N4 (we ignore it)
 		// Server's port for data connection is P1*256+P2
-		str = strrchr(buf, ')');
-		if (str) str[0] = '\0';
-		str = strrchr(buf, ',');
-		if (!str) goto pasv_error;
-		port = xatou_range(str+1, 0, 255);
-		*str = '\0';
-		str = strrchr(buf, ',');
-		if (!str) goto pasv_error;
-		port += xatou_range(str+1, 0, 255) * 256;
-		set_nport(lsa, htons(port));
+		if (!(str = strrchr(buf, ')')))
+ 			goto pasv_error;
+ 		*str = 0;	// clear trailing '\0'
+ 		if (!(str = strrchr(buf, '(')))
+ 			goto pasv_error;
+ 		p = ++str;
+ 		for (nn=0; nn<4; nn++) {
+ 			if (!(str = strchr(str, ',')))
+ 				goto pasv_error;
+ 			*str++ = '.';
+ 		}
+ 		str[-1] = 0;
+ 		if (inet_aton(p, &lsa->u.sin.sin_addr) == 0)
+ 			goto pasv_error;
+ 		if (!(str = strchr((p=str), ',')))
+ 			goto pasv_error;
+ 		*str++ = 0;
+ 		set_nport(lsa, htons(xatou_range(p, 0, 255) * 256 + xatou_range(str, 0, 255)));
 		dfp = open_socket(lsa);
 
 		if (beg_range) {
