--- busybox-1.00-pre7/init/init.c	Fri Sep 26 12:45:55 2003
+++ busybox-1.00-pre7-p/init/init.c	Fri Feb  6 09:41:23 2004
@@ -110,6 +110,7 @@
 #define KERNEL_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))
 
 #define INITTAB      "/etc/inittab"	/* inittab file location */
+#define INITTAB_FB	 "/usr/share/defaults/inittab" /* fallback inittab */
 #ifndef INIT_SCRIPT
 #define INIT_SCRIPT  "/etc/init.d/rcS"	/* Default sysinit script. */
 #endif
@@ -949,7 +950,8 @@
 	const struct init_action_type *a = actions;
 
 
-	file = fopen(INITTAB, "r");
+	if ( (file = fopen(INITTAB, "r")) == NULL )
+		file = fopen(INITTAB_FB,"r");
 	if (file == NULL) {
 		/* No inittab file -- set up some default behavior */
 #endif
