--- busybox-1.00-pre7/editors/vi.c	Wed Feb  4 12:19:44 2004
+++ busybox-1.00-pre7-p/editors/vi.c	Fri Feb  6 09:41:41 2004
@@ -1672,7 +1672,7 @@
  		// if we are at the next word's first char
  		// step back one char
  		// but check the possibilities when it is true
- 		if (dot > text && ((isspace(dot[0]) && !isspace(dot[0]))
+ 		if (dot > text && ((isspace(dot[-1]) && !isspace(dot[0]))
  				|| (ispunct(dot[-1]) && !ispunct(dot[0]))
  				|| (isalnum(dot[-1]) && !isalnum(dot[0]))))
   			dot--;		// move back off of next word
