--- busybox-1.10.4/init/init.c.orig	2008-06-25 12:55:32.000000000 +0200
+++ busybox-1.10.4/init/init.c	2008-07-08 19:13:18.000000000 +0200
@@ -28,6 +28,7 @@
 #include <sys/resource.h>
 
 #define INITTAB      "/etc/inittab"	/* inittab file location */
+#define INITTAB_FB	 "/usr/share/defaults/etc/inittab" /* fallback inittab */
 #ifndef INIT_SCRIPT
 #define INIT_SCRIPT  "/etc/init.d/rcS"	/* Default sysinit script. */
 #endif
@@ -703,10 +704,12 @@
 	FILE *file;
 	char buf[COMMAND_SIZE];
 
-	if (ENABLE_FEATURE_USE_INITTAB)
-		file = fopen(INITTAB, "r");
-	else
+	if (ENABLE_FEATURE_USE_INITTAB) {
+		if ( (file = fopen(INITTAB, "r")) == NULL )
+			file = fopen(INITTAB_FB, "r");
+	} else {
 		file = NULL;
+	}
 
 	/* No inittab file -- set up some default behavior */
 	if (file == NULL) {
