--- busybox-1.19.3/util-linux/rdate.c.orig	2011-09-06 04:35:17.000000000 +0200
+++ busybox-1.19.3/util-linux/rdate.c	2011-12-07 14:18:57.351679473 +0100
@@ -9,11 +9,12 @@
 */
 
 //usage:#define rdate_trivial_usage
-//usage:       "[-sp] HOST"
+//usage:       "[-spa] HOST"
 //usage:#define rdate_full_usage "\n\n"
 //usage:       "Get and possibly set the system date and time from a remote HOST\n"
 //usage:     "\n	-s	Set the system date and time (default)"
 //usage:     "\n	-p	Print the date and time"
+//usage:     "\n	-a	Skew the time instead of just hopping it"
 
 #include "libbb.h"
 
@@ -56,7 +57,7 @@
 	unsigned long flags;
 
 	opt_complementary = "-1";
-	flags = getopt32(argv, "sp");
+	flags = getopt32(argv, "spa");
 
 	remote_time = askremotedate(argv[optind]);
 
@@ -67,6 +68,11 @@
 		if (current_time == remote_time)
 			bb_error_msg("current time matches remote time");
 		else
+		if (flags & 4) {
+			struct timeval tv = { .tv_sec = remote_time - current_time, .tv_usec = 0 };
+			if (adjtime(&tv, NULL) == -1)
+				bb_perror_msg_and_die("cannot adjust time of day");
+		} else
 			if (stime(&remote_time) < 0)
 				bb_perror_msg_and_die("can't set time of day");
 	}
