--- busybox-1.00-pre9/networking/wget.c.orig	Tue Apr 13 18:18:16 2004
+++ busybox-1.00-pre9/networking/wget.c	Tue Apr 13 18:25:49 2004
@@ -674,6 +674,13 @@
 			bb_perror_msg_and_die("fgets()");
 	} while (! isdigit(buf[0]) || buf[3] != ' ');
 
+	// remove the trailing '\r' and '\n' chars
+	// from the response
+	p=buf+strlen(buf)-1;
+	while( p >= buf && ( *p == '\r' || *p == '\n' ) )
+		p--;
+	p[1]=0;
+
 	return atoi(buf);
 }
 
