--- busybox-1.01/modutils/modprobe.c.orig	2005-11-20 17:36:00.000000000 +0100
+++ busybox-1.01/modutils/modprobe.c	2005-11-20 17:36:37.000000000 +0100
@@ -569,7 +569,7 @@
 	check_dep ( mod, &head, &tail );
 
 	if ( head && tail ) {
-#if defined(CONFIG_FEATURE_2_6_MODULES)
+//#if defined(CONFIG_FEATURE_2_6_MODULES)
 		if ( argc ) {
 			int i;
 			int l = 0;
@@ -586,7 +586,7 @@
 				strcat ( head-> m_options, " " );
 			}
 		}
-#endif
+//#endif
 
 		// process tail ---> head
 		rc = mod_process ( tail, 1 );
