--- busybox/procps/ps.c.orig	2007-11-10 02:40:53.000000000 +0100
+++ busybox/procps/ps.c	2007-11-20 15:15:10.000000000 +0100
@@ -357,7 +357,7 @@
 #if ENABLE_FEATURE_PS_WIDE || ENABLE_SELINUX
 #if ENABLE_FEATURE_PS_WIDE
 	opt_complementary = "-:ww";
-	USE_SELINUX(i =) getopt32(argv, USE_SELINUX("Z") "w", &w_count);
+	USE_SELINUX(i =) getopt32(argv, USE_SELINUX("Z") "wafnrstuvx", &w_count);
 	/* if w is given once, GNU ps sets the width to 132,
 	 * if w is given more than once, it is "unlimited"
 	 */
@@ -370,7 +370,7 @@
 			terminal_width = MAX_WIDTH;
 	}
 #else /* only ENABLE_SELINUX */
-	i = getopt32(argv, "Z");
+	i = getopt32(argv, "Zafnrstuvx");
 #endif
 #if ENABLE_SELINUX
 	if ((i & 1) && is_selinux_enabled())
