--- busybox/networking/wget.c.orig	2007-11-20 14:59:03.000000000 +0100
+++ busybox/networking/wget.c	2007-11-20 15:00:54.000000000 +0100
@@ -754,7 +754,7 @@
 
 	/* Do it before progressmeter (want to have nice error message) */
 	if (output_fd < 0) {
-		int o_flags = O_WRONLY | O_CREAT | O_TRUNC | O_EXCL;
+		int o_flags = O_WRONLY | O_CREAT | O_TRUNC;
 		/* compat with wget: -O FILE can overwrite */
 		if (opt & WGET_OPT_OUTNAME)
 			o_flags = O_WRONLY | O_CREAT | O_TRUNC;
