--- busybox-1.1.2/procps/ps.c.orig	2006-05-14 09:14:54.000000000 +0200
+++ busybox-1.1.2/procps/ps.c	2006-05-14 09:15:04.000000000 +0200
@@ -44,11 +44,11 @@
 #if ENABLE_FEATURE_PS_WIDE || ENABLE_SELINUX
 	/* handle arguments */
 #if ENABLE_FEATURE_PS_WIDE && ENABLE_SELINUX
-	i = bb_getopt_ulflags(argc, argv, "wc", &w_count);
+	i = bb_getopt_ulflags(argc, argv, "wcx", &w_count);
 #elif ENABLE_FEATURE_PS_WIDE && !ENABLE_SELINUX
-	bb_getopt_ulflags(argc, argv, "w", &w_count);
+	bb_getopt_ulflags(argc, argv, "wax", &w_count);
 #else /* !ENABLE_FEATURE_PS_WIDE && ENABLE_SELINUX */
-	i = bb_getopt_ulflags(argc, argv, "c");
+	i = bb_getopt_ulflags(argc, argv, "cax");
 #endif
 #if ENABLE_FEATURE_PS_WIDE
 	/* if w is given once, GNU ps sets the width to 132,
