--- busybox-1.00-pre7/shell/ash.c	Wed Feb  4 09:24:39 2004
+++ busybox-1.00-pre7-p/shell/ash.c	Fri Feb  6 09:41:23 2004
@@ -4126,9 +4126,6 @@
 		firstchange = 0;
 	clearcmdentry(firstchange);
 	builtinloc = idx_bltin;
-#ifdef CONFIG_FEATURE_COMMAND_TAB_COMPLETION
-	cmdedit_path_lookup = newval;
-#endif
 }
 
 
@@ -6026,6 +6023,7 @@
 	if (!iflag || parsefile->fd)
 		nr = safe_read(parsefile->fd, buf, BUFSIZ - 1);
 	else {
+		cmdedit_path_lookup = pathval();
 		nr = cmdedit_read_input((char *) cmdedit_prompt, buf);
 		if(nr == 0) {
 			/* Ctrl+C presend */
