--- freeswan-2.05/programs/_plutorun/_plutorun.in.orig	Sat Apr 10 23:15:43 2004
+++ freeswan-2.05/programs/_plutorun/_plutorun.in	Sat Apr 10 23:16:37 2004
@@ -155,6 +155,7 @@
 echo "$verb Pluto subsystem..." | logger -p authpriv.error -t ipsec__plutorun
 execdir=${IPSEC_EXECDIR-@IPSEC_EXECDIR@}
 libdir=${IPSEC_LIBDIR-@IPSEC_LIBDIR@}
+exec 2>&-
 until (
 	if test -s $info
 	then
--- freeswan-2.05/programs/showhostkey/showhostkey.in.orig	Tue Apr 13 16:31:32 2004
+++ freeswan-2.05/programs/showhostkey/showhostkey.in	Tue Apr 13 16:31:48 2004
@@ -70,7 +70,7 @@
 	exit 1
 fi
 
-host="`hostname --fqdn`"
+host="`hostname -f`"
 
 awk '	BEGIN {
 		inkey = 0
