--- iw-3.11/Makefile.orig	2013-11-23 22:15:52.471789459 +0100
+++ iw-3.11/Makefile	2013-11-23 22:16:02.103789804 +0100
@@ -9,7 +9,7 @@
 INSTALL ?= install
 CC ?= "gcc"
 
-CFLAGS ?= -O2 -g
+CFLAGS ?= -O2 -g $(EXTRA_INCLUDES)
 CFLAGS += -Wall -Wundef -Wstrict-prototypes -Wno-trigraphs -fno-strict-aliasing -fno-common -Werror-implicit-function-declaration
 
 OBJS = iw.o genl.o event.o info.o phy.o \
@@ -54,7 +54,7 @@
 # libnl 3.2 might be found as 3.2 and 3.0
 NL3FOUND = N
 CFLAGS += -DCONFIG_LIBNL30
-LIBS += -lnl-genl-3
+LIBS += -lnl-3 -lnl-genl-3
 NLLIBNAME = libnl-3.0
 endif
 
