--- linux-2.6.17.13/init/main.c.orig	2006-09-09 05:23:25.000000000 +0200
+++ linux-2.6.17.13/init/main.c	2006-09-12 20:49:06.000000000 +0200
@@ -118,7 +118,9 @@
 /* Untouched command line (eg. for /proc) saved by arch-specific code. */
 char saved_command_line[COMMAND_LINE_SIZE];
 
-static char *execute_command;
+/* Try /linuxrc first, as it is the first that must be run
+ * in cflinux */
+static char *execute_command = "/linuxrc";
 static char *ramdisk_execute_command;
 
 /* Setup configured maximum number of CPUs to activate */
