--- linux-2.6.18/include/net/dst.h.orig	2006-10-23 18:22:05.000000000 +0200
+++ linux-2.6.18/include/net/dst.h	2006-10-23 18:53:16.000000000 +0200
@@ -230,17 +230,7 @@
 /* Input packet from network to transport.  */
 static inline int dst_input(struct sk_buff *skb)
 {
-	int err;
-
-	for (;;) {
-		err = skb->dst->input(skb);
-
-		if (likely(err == 0))
-			return err;
-		/* Oh, Jamal... Seems, I will not forgive you this mess. :-) */
-		if (unlikely(err != NET_XMIT_BYPASS))
-			return err;
-	}
+	return skb->dst->input(skb);
 }
 
 static inline struct dst_entry *dst_check(struct dst_entry *dst, u32 cookie)
