--- linux-2.6.20.3/include/net/ip_fib.h.orig	2007-04-09 10:59:51.000000000 +0200
+++ linux-2.6.20.3/include/net/ip_fib.h	2007-04-09 13:22:00.000000000 +0200
@@ -179,14 +179,23 @@
 
 #ifndef CONFIG_IP_MULTIPLE_TABLES
 
-extern struct fib_table *ip_fib_local_table;
 extern struct fib_table *ip_fib_main_table;
 
+#ifdef CONFIG_IP_FIB_HASH
+#define _tb_lookup fn_hash_lookup
+#endif
+#ifdef CONFIG_IP_FIB_TRIE
+#define _tb_lookup fn_trie_lookup
+#endif
+#ifdef CONFIG_IP_FIB_LEF
+#define _tb_lookup fn_lef_lookup
+#endif
+
+extern int _tb_lookup(struct fib_table *, const struct flowi *, struct fib_result *);
+
 static inline struct fib_table *fib_get_table(u32 id)
 {
-	if (id != RT_TABLE_LOCAL)
-		return ip_fib_main_table;
-	return ip_fib_local_table;
+	return ip_fib_main_table;
 }
 
 static inline struct fib_table *fib_new_table(u32 id)
@@ -196,10 +205,7 @@
 
 static inline int fib_lookup(const struct flowi *flp, struct fib_result *res)
 {
-	if (ip_fib_local_table->tb_lookup(ip_fib_local_table, flp, res) &&
-	    ip_fib_main_table->tb_lookup(ip_fib_main_table, flp, res))
-		return -ENETUNREACH;
-	return 0;
+	return _tb_lookup(ip_fib_main_table, flp, res);
 }
 
 static inline void fib_select_default(const struct flowi *flp, struct fib_result *res)
--- linux-2.6.18.orig/net/ipv4/fib_hash.c	2006-09-20 05:42:06.000000000 +0200
+++ linux-2.6.18/net/ipv4/fib_hash.c	2006-10-22 13:37:29.000000000 +0200
@@ -242,7 +242,10 @@
 	return fz;
 }
 
-static int
+#ifdef CONFIG_IP_MULTIPLE_TABLES
+static
+#endif
+int
 fn_hash_lookup(struct fib_table *tb, const struct flowi *flp, struct fib_result *res)
 {
 	int err;
--- linux-2.6.18.orig/net/ipv4/fib_trie.c	2006-09-20 05:42:06.000000000 +0200
+++ linux-2.6.18/net/ipv4/fib_trie.c	2006-10-22 13:37:52.000000000 +0200
@@ -1306,7 +1306,10 @@
 	return 1;
 }
 
-static int
+#ifdef CONFIG_IP_MULTIPLE_TABLES
+static
+#endif
+int
 fn_trie_lookup(struct fib_table *tb, const struct flowi *flp, struct fib_result *res)
 {
 	struct trie *t = (struct trie *) tb->tb_data;
--- linux-2.6.20.3/net/ipv4/fib_lef.c.orig	2007-04-09 13:23:22.000000000 +0200
+++ linux-2.6.20.3/net/ipv4/fib_lef.c	2007-04-09 13:24:44.000000000 +0200
@@ -333,7 +333,10 @@
 	return err;
 }
 
-static int
+#ifdef CONFIG_IP_MULTIPLE_TABLES
+static
+#endif
+int
 fn_lef_lookup(struct fib_table *tb, const struct flowi *flp, struct fib_result *res)
 {
 	lef_t			*ln;		// current lef node
--- linux-2.6.19/net/ipv4/fib_frontend.c.orig	2006-12-03 10:36:29.000000000 +0100
+++ linux-2.6.19/net/ipv4/fib_frontend.c	2006-12-03 10:41:17.000000000 +0100
@@ -52,7 +52,6 @@
 
 #ifndef CONFIG_IP_MULTIPLE_TABLES
 
-struct fib_table *ip_fib_local_table;
 struct fib_table *ip_fib_main_table;
 
 #define FIB_TABLE_HASHSZ 1
@@ -132,8 +131,12 @@
 	res.r = NULL;
 #endif
 
+#ifdef CONFIG_IP_MULTIPLE_TABLES
 	if (!ip_fib_local_table ||
 	    ip_fib_local_table->tb_lookup(ip_fib_local_table, &fl, &res))
+#else
+	if (ip_fib_main_table->tb_lookup(ip_fib_main_table, &fl, &res))
+#endif
 		return NULL;
 	if (res.type != RTN_LOCAL)
 		goto out;
@@ -161,14 +164,20 @@
 	res.r = NULL;
 #endif
 	
+#ifdef CONFIG_IP_MULTIPLE_TABLES
 	if (ip_fib_local_table) {
 		ret = RTN_UNICAST;
 		if (!ip_fib_local_table->tb_lookup(ip_fib_local_table,
 						   &fl, &res)) {
+#else
+	if (!ip_fib_main_table->tb_lookup(ip_fib_main_table, &fl, &res)) {
+#endif
 			ret = res.type;
 			fib_res_put(&res);
 		}
+#ifdef CONFIG_IP_MULTIPLE_TABLES
 	}
+#endif
 	return ret;
 }
 
@@ -641,6 +650,7 @@
 		.fc_nlflags = NLM_F_CREATE | NLM_F_APPEND,
 	};
 
+#ifdef CONFIG_IP_MULTIPLE_TABLES
 	if (type == RTN_UNICAST)
 		tb = fib_new_table(RT_TABLE_MAIN);
 	else
@@ -648,6 +658,9 @@
 
 	if (tb == NULL)
 		return;
+#else
+	tb = ip_fib_main_table;
+#endif
 
 	cfg.fc_table = tb->tb_id;
 
@@ -911,8 +924,6 @@
 	for (i = 0; i < FIB_TABLE_HASHSZ; i++)
 		INIT_HLIST_HEAD(&fib_table_hash[i]);
 #ifndef CONFIG_IP_MULTIPLE_TABLES
-	ip_fib_local_table = fib_hash_init(RT_TABLE_LOCAL);
-	hlist_add_head_rcu(&ip_fib_local_table->tb_hlist, &fib_table_hash[0]);
 	ip_fib_main_table  = fib_hash_init(RT_TABLE_MAIN);
 	hlist_add_head_rcu(&ip_fib_main_table->tb_hlist, &fib_table_hash[0]);
 #else
