--- mstpd-code-r34/Makefile.orig	2012-08-26 08:44:20.565929618 +0200
+++ mstpd-code-r34/Makefile	2012-08-26 08:44:29.441930283 +0200
@@ -34,7 +34,7 @@
 	-mkdir -pv $(DESTDIR)/sbin
 	install -m 755 mstpd $(DESTDIR)/sbin/mstpd
 	install -m 755 mstpctl $(DESTDIR)/sbin/mstpctl
-	install -m 755 bridge-stp /sbin/bridge-stp
+	install -m 755 bridge-stp $(DESTDIR)/sbin/bridge-stp
 
 romfs: all
 	$(ROMFSINST) /sbin/mstpd
