--- net-snmp-5.2.3/configure.orig	2009-11-14 16:15:59.000000000 +0000
+++ net-snmp-5.2.3/configure	2009-11-14 16:18:32.000000000 +0000
@@ -4940,7 +4940,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   case $host_cpu in
   alpha* | hppa* | i*86 | mips | mipsel | powerpc* | sparc* | ia64*)
     lt_cv_deplibs_check_method=pass_all ;;
@@ -7709,7 +7709,7 @@
   ;;
 
 # This must be Linux ELF.
-linux-gnu*)
+linux-gnu*|linux-uclibc*)
   version_type=linux
   need_lib_prefix=no
   need_version=no
