--- openswan-2.4.5/programs/_realsetup/_realsetup.in.orig	2006-05-09 22:05:56.000000000 +0200
+++ openswan-2.4.5/programs/_realsetup/_realsetup.in	2006-05-09 23:34:15.000000000 +0200
@@ -16,6 +16,9 @@
 
 me='ipsec setup'		# for messages
 
+### this is to have a ps behaving as needed - cflinux hack
+[ -r /usr/lib/cflinux/hackps.sh ] && . /usr/lib/cflinux/hackps.sh
+
 # Misc. paths (some of this should perhaps be overrideable from ipsec.conf).
 plutopid=/var/run/pluto/pluto.pid
 plutoctl=/var/run/pluto/pluto.ctl
