--- openswan-2.4.5/programs/_realsetup/_realsetup.in.orig	2006-05-10 18:38:46.000000000 +0200
+++ openswan-2.4.5/programs/_realsetup/_realsetup.in	2006-05-10 18:39:05.000000000 +0200
@@ -147,8 +147,8 @@
 	done ;
 	test -r /proc/net/ipsec_klipsdebug && ipsec klipsdebug --none ;
 	test -d /proc/net/ipsec/eroute && ipsec eroute --clear ;
-	test -d /proc/net/ipsec/spi && ipsec spi --clear'
-	#lsmod 2>&1 | grep "^ipsec" > /dev/null && rmmod ipsec'
+	test -d /proc/net/ipsec/spi && ipsec spi --clear ;
+	lsmod 2>&1 | grep "^ipsec" > /dev/null && rmmod ipsec'
 fi
 
 if $netkey
