--- openswan-2.2.0/programs/setup/Makefile.orig	2006-05-09 10:07:44.000000000 +0200
+++ openswan-2.2.0/programs/setup/Makefile	2006-05-09 10:10:03.000000000 +0200
@@ -31,16 +31,16 @@
 # the priorities match those in setup's chkconfig line
 install:: setup
 	@rm -f $(BINDIR)/setup
-	@$(INSTALL) $(INSTBINFLAGS) setup $(RCDIR)/ipsec
-	@ln -s $(FINALRCDIR)/ipsec $(BINDIR)/setup
-	-@for i in 0 1 2 3 4 5 6; do mkdir -p $(RCDIR)/../rc$$i.d; done
-	-@cd $(RCDIR)/../rc0.d && ln -f -s ../init.d/ipsec K76ipsec
-	-@cd $(RCDIR)/../rc1.d && ln -f -s ../init.d/ipsec K76ipsec
-	-@cd $(RCDIR)/../rc2.d && ln -f -s ../init.d/ipsec S47ipsec
-	-@cd $(RCDIR)/../rc3.d && ln -f -s ../init.d/ipsec S47ipsec
-	-@cd $(RCDIR)/../rc4.d && ln -f -s ../init.d/ipsec S47ipsec
-	-@cd $(RCDIR)/../rc5.d && ln -f -s ../init.d/ipsec S47ipsec
-	-@cd $(RCDIR)/../rc6.d && ln -f -s ../init.d/ipsec K76ipsec
+	@$(INSTALL) $(INSTBINFLAGS) setup $(BINDIR)/
+#	@ln -s $(FINALRCDIR)/ipsec $(BINDIR)/setup
+#	-@for i in 0 1 2 3 4 5 6; do mkdir -p $(RCDIR)/../rc$$i.d; done
+#	-@cd $(RCDIR)/../rc0.d && ln -f -s ../init.d/ipsec K76ipsec
+#	-@cd $(RCDIR)/../rc1.d && ln -f -s ../init.d/ipsec K76ipsec
+#	-@cd $(RCDIR)/../rc2.d && ln -f -s ../init.d/ipsec S47ipsec
+#	-@cd $(RCDIR)/../rc3.d && ln -f -s ../init.d/ipsec S47ipsec
+#	-@cd $(RCDIR)/../rc4.d && ln -f -s ../init.d/ipsec S47ipsec
+#	-@cd $(RCDIR)/../rc5.d && ln -f -s ../init.d/ipsec S47ipsec
+#	-@cd $(RCDIR)/../rc6.d && ln -f -s ../init.d/ipsec K76ipsec
 
 install_file_list::
 	@echo $(RCDIR)/ipsec
