--- perl-5.8.7.orig/Configure	2005-04-20 18:34:37.000000000 +0200
+++ perl-5.8.7/Configure	2005-07-26 19:27:10.000000000 +0200
@@ -907,7 +907,7 @@
 longsize=''
 shortsize=''
 issymlink=''
-libc=''
+libc="$UC_ROOT/lib/libuClibc-0.9.26.so"
 ldlibpthname=''
 libperl=''
 shrpenv=''
@@ -1225,8 +1225,9 @@
 archname=''
 : Possible local include directories to search.
 : Set locincpth to "" in a hint file to defeat local include searches.
-locincpth="/usr/local/include /opt/local/include /usr/gnu/include"
-locincpth="$locincpth /opt/gnu/include /usr/GNU/include /opt/GNU/include"
+locincpth=""
+#locincpth="/usr/local/include /opt/local/include /usr/gnu/include"
+#locincpth="$locincpth /opt/gnu/include /usr/GNU/include /opt/GNU/include"
 :
 : no include file wanted by default
 inclwanted=''
@@ -1236,14 +1237,14 @@
 : change the next line if compiling for Xenix/286 on Xenix/386
 xlibpth='/usr/lib/386 /lib/386'
 : Possible local library directories to search.
-loclibpth="/usr/local/lib /opt/local/lib /usr/gnu/lib"
-loclibpth="$loclibpth /opt/gnu/lib /usr/GNU/lib /opt/GNU/lib"
+#loclibpth="/usr/local/lib /opt/local/lib /usr/gnu/lib"
+#loclibpth="$loclibpth /opt/gnu/lib /usr/GNU/lib /opt/GNU/lib"
 
 : general looking path for locating libraries
-glibpth="/lib /usr/lib $xlibpth"
-glibpth="$glibpth /usr/ccs/lib /usr/ucblib /usr/local/lib"
-test -f /usr/shlib/libc.so && glibpth="/usr/shlib $glibpth"
-test -f /shlib/libc.so     && glibpth="/shlib $glibpth"
+glibpth="$UC_ROOT/lib"
+#glibpth="$glibpth /usr/ccs/lib /usr/ucblib /usr/local/lib"
+#test -f /usr/shlib/libc.so && glibpth="/usr/shlib $glibpth"
+#test -f /shlib/libc.so     && glibpth="/shlib $glibpth"
 
 : Private path used by Configure to find libraries.  Its value
 : is prepended to libpth. This variable takes care of special
--- perl-5.8.7.orig/hints/linux.sh	2005-04-04 22:08:31.000000000 +0200
+++ perl-5.8.7/hints/linux.sh	2005-07-26 22:09:40.000000000 +0200
@@ -52,10 +52,10 @@
 # We don't use __GLIBC__ and  __GLIBC_MINOR__ because they
 # are insufficiently precise to distinguish things like
 # libc-2.0.6 and libc-2.0.7.
-if test -L /lib/libc.so.6; then
-    libc=`ls -l /lib/libc.so.6 | awk '{print $NF}'`
-    libc=/lib/$libc
-fi
+#if test -L /lib/libc.so.6; then
+#    libc=`ls -l /lib/libc.so.6 | awk '{print $NF}'`
+#    libc=/lib/$libc
+#fi
 
 # Configure may fail to find lstat() since it's a static/inline
 # function in <sys/stat.h>.
@@ -315,3 +315,5 @@
 	;;
 esac
 EOCBU
+
+ccflags="$ccflags -DPERL_USE_SAFE_PUTENV"
