--- ppp-cvs-20050725-p1/pppd/plugins/radius/radius.c.orig	2007-05-21 09:56:10.000000000 +0200
+++ ppp-cvs-20050725-p1/pppd/plugins/radius/radius.c	2007-05-21 10:30:19.000000000 +0200
@@ -942,6 +942,10 @@
 	return;
     }
 
+    if (rstate.acct_interim_interval)  {
+	    UNTIMEOUT(radius_acct_interim, NULL);
+    }
+
     rstate.accounting_started = 0;
     rc_avpair_add(&send, PW_ACCT_SESSION_ID, rstate.session_id,
 		   0, VENDOR_NONE);
