--- quagga-0.96.5/configure.orig	Thu Nov 18 11:25:18 2004
+++ quagga-0.96.5/configure	Thu Nov 18 11:46:44 2004
@@ -7546,7 +7546,7 @@
 if test "${enable_snmp}" = "yes";then
   if test "${HAVE_SNMP}" = ""; then
     old_libs="${LIBS}"
-    LIBS="-L/usr/lib"
+    LIBS=""
     unset ac_cv_lib_snmp_asn_parse_int
     echo "$as_me:$LINENO: checking for main in -lcrypto" >&5
 echo $ECHO_N "checking for main in -lcrypto... $ECHO_C" >&6
@@ -7723,7 +7723,7 @@
   fi
   if test "${HAVE_SNMP}" = ""; then
     old_libs="${LIBS}"
-    LIBS="-L/usr/lib"
+    LIBS=""
     unset ac_cv_lib_snmp_asn_parse_int
     echo "$as_me:$LINENO: checking for asn_parse_int in -lsnmp" >&5
 echo $ECHO_N "checking for asn_parse_int in -lsnmp... $ECHO_C" >&6
@@ -8082,13 +8082,13 @@
   fi
 
   if test "${HAVE_SNMP}" = "yes"; then
-    for ac_snmp in /usr/include/net-snmp/library/asn1.h /usr/include/ucd-snmp/asn1.h /usr/local/include/ucd-snmp/asn1.h /dev/null
+    for ac_snmp in ${UC_ROOT}/include/net-snmp/library/asn1.h /usr/include/net-snmp/library/asn1.h /usr/include/ucd-snmp/asn1.h /usr/local/include/ucd-snmp/asn1.h /dev/null
     do
       test -f "${ac_snmp}" && break
     done
 
     case ${ac_snmp} in
-      /usr/include/net-snmp/*)
+      */include/net-snmp/*)
 
 cat >>confdefs.h <<\_ACEOF
 #define HAVE_SNMP
@@ -8104,50 +8104,24 @@
 #define UCD_COMPATIBLE
 _ACEOF
 
-                  CFLAGS="${CFLAGS} -I/usr/include/net-snmp -I/usr/include/net-snmp/library"
+                  ac_snmp_inc_base=$(echo ${ac_snmp}|sed 's/^\(.*\/include\/net-snmp\).*$/\1/')
+                  CFLAGS="${CFLAGS} -I${ac_snmp_inc_base} -I${ac_snmp_inc_base}/library"
 		  if test "${HAVE_NETSNMP}" = "yes"; then
 		        LIBS="${LIBS} -lnetsnmp"
 		  else
 			LIBS="${LIBS} -lsnmp"
 		  fi
                   ;;
-      /usr/include/ucd-snmp/*)
+      */include/ucd-snmp/*)
 
 cat >>confdefs.h <<\_ACEOF
 #define HAVE_SNMP
 _ACEOF
 
-                  CFLAGS="${CFLAGS} -I/usr/include/ucd-snmp"
+                  ac_snmp_inc_base=$(echo ${ac_snmp}|sed 's/^\(.*\/include\/ucd-snmp\).*$/\1/')
+                  CFLAGS="${CFLAGS} -I${ac_snmp_inc_base}"
     		  LIBS="${LIBS} -lsnmp"
                   ;;
-      /usr/local/include/ucd-snmp/*)
-
-cat >>confdefs.h <<\_ACEOF
-#define HAVE_SNMP
-_ACEOF
-
-  	          CFLAGS="${CFLAGS} -I/usr/local/include/ucd-snmp"
-  		  LIBS="${LIBS} -L/usr/local/lib -lsnmp"
-                  ;;
-      /usr/local/include/net-snmp/*)
-
-cat >>confdefs.h <<\_ACEOF
-#define HAVE_SNMP
-_ACEOF
-
-
-cat >>confdefs.h <<\_ACEOF
-#define HAVE_NET_SNMP
-_ACEOF
-
-
-cat >>confdefs.h <<\_ACEOF
-#define UCD_COMPATIBLE
-_ACEOF
-
-      		CFLAGS="${CFLAGS} -I/usr/local/include/net-snmp"
-      		LIBS="${LIBS} -L/usr/local/lib -lnetsnmp"
-      		;;
     esac
     if test "${NEED_CRYPTO}" = "yes"; then
       LIBS="${LIBS} -lcrypto"
--- quagga-0.98.4/ospfd/Makefile.in.orig	Mon Jun 27 19:05:41 2005
+++ quagga-0.98.4/ospfd/Makefile.in	Tue Jul 26 09:54:19 2005
@@ -250,7 +250,7 @@
 	ospf_flood.h ospf_snmp.h ospf_te.h ospf_vty.h ospf_apiserver.h
 
 ospfd_SOURCES = ospf_main.c
-ospfd_LDADD = libospf.la ../lib/libzebra.la @LIBCAP@
+ospfd_LDADD = .libs/libospf.a ../lib/libzebra.la @LIBCAP@
 EXTRA_DIST = OSPF-MIB.txt OSPF-TRAP-MIB.txt ChangeLog.opaque.txt
 examplesdir = $(exampledir)
 dist_examples_DATA = ospfd.conf.sample
--- quagga-0.98.6.orig/lib/zebra.h	2005-06-15 13:54:18.000000000 +0200
+++ quagga-0.98.6/lib/zebra.h	2006-12-03 19:50:26.000000000 +0100
@@ -152,6 +152,9 @@
 #ifdef HAVE_NETLINK
 #include <linux/netlink.h>
 #include <linux/rtnetlink.h>
+#include <linux/if_addr.h>
+#define IFA_RTA(r)  ((struct rtattr*)(((char*)(r)) + NLMSG_ALIGN(sizeof(struct ifaddrmsg))))
+#define IFLA_RTA(r)  ((struct rtattr*)(((char*)(r)) + NLMSG_ALIGN(sizeof(struct ifinfomsg))))
 #else
 #define RT_TABLE_MAIN		0
 #endif /* HAVE_NETLINK */
