--- quagga-0.98.4/ospfd/Makefile.in.orig	Mon Jun 27 19:05:41 2005
+++ quagga-0.98.4/ospfd/Makefile.in	Tue Jul 26 09:54:19 2005
@@ -250,7 +250,7 @@
 	ospf_flood.h ospf_snmp.h ospf_te.h ospf_vty.h ospf_apiserver.h
 
 ospfd_SOURCES = ospf_main.c
-ospfd_LDADD = libospf.la ../lib/libzebra.la @LIBCAP@
+ospfd_LDADD = .libs/libospf.a ../lib/libzebra.la @LIBCAP@
 EXTRA_DIST = OSPF-MIB.txt OSPF-TRAP-MIB.txt ChangeLog.opaque.txt
 examplesdir = $(exampledir)
 dist_examples_DATA = ospfd.conf.sample
