diff -ruNP quagga-0.96.5/bgpd/Makefile.in quagga-0.96.5-patched/bgpd/Makefile.in
--- quagga-0.96.5/bgpd/Makefile.in	Mon May  3 15:03:32 2004
+++ quagga-0.96.5-patched/bgpd/Makefile.in	Wed May  5 16:07:00 2004
@@ -168,7 +168,7 @@
 	bgp_main.c $(libbgp_a_SOURCES)
 
 
-bgpd_LDADD = ../lib/libzebra.a @LIBCAP@
+bgpd_LDADD = -L../lib -lzebra @LIBCAP@
 
 examplesdir = $(exampledir)
 dist_examples_DATA = bgpd.conf.sample bgpd.conf.sample2
@@ -207,7 +207,7 @@
 	bgp_advertise.$(OBJEXT) bgp_vty.$(OBJEXT)
 am_bgpd_OBJECTS = bgp_main.$(OBJEXT) $(am__objects_1)
 bgpd_OBJECTS = $(am_bgpd_OBJECTS)
-bgpd_DEPENDENCIES = ../lib/libzebra.a
+bgpd_DEPENDENCIES =
 bgpd_LDFLAGS =
 
 DEFAULT_INCLUDES =  -I. -I$(srcdir) -I$(top_builddir)
diff -ruNP quagga-0.96.5/isisd/Makefile.in quagga-0.96.5-patched/isisd/Makefile.in
--- quagga-0.96.5/isisd/Makefile.in	Mon May  3 15:07:05 2004
+++ quagga-0.96.5-patched/isisd/Makefile.in	Wed May  5 16:08:49 2004
@@ -167,7 +167,7 @@
 	isis_main.c $(libisis_a_SOURCES)
 
 
-isisd_LDADD = ../lib/libzebra.a @LIBCAP@
+isisd_LDADD = -L../lib -lzebra @LIBCAP@
 
 examplesdir = $(exampledir)
 dist_examples_DATA = isisd.conf.sample
@@ -200,7 +200,7 @@
 	isis_spf.$(OBJEXT) isis_route.$(OBJEXT) isis_routemap.$(OBJEXT)
 am_isisd_OBJECTS = isis_main.$(OBJEXT) $(am__objects_1)
 isisd_OBJECTS = $(am_isisd_OBJECTS)
-isisd_DEPENDENCIES = ../lib/libzebra.a
+isisd_DEPENDENCIES =
 isisd_LDFLAGS =
 
 DEFAULT_INCLUDES =  -I. -I$(srcdir) -I$(top_builddir)
diff -ruNP quagga-0.96.5/lib/Makefile.in quagga-0.96.5-patched/lib/Makefile.in
--- quagga-0.96.5/lib/Makefile.in	Mon May  3 15:02:36 2004
+++ quagga-0.96.5-patched/lib/Makefile.in	Wed May  5 16:07:00 2004
@@ -144,9 +144,9 @@
 sysconfdir = @sysconfdir@
 target_alias = @target_alias@
 
-lib_LIBRARIES = libzebra.a
+lib_LIBRARIES = libzebra.so
 
-libzebra_a_SOURCES = \
+libzebra_so_SOURCES = \
 	version.c network.c pid_output.c getopt.c getopt1.c daemon.c \
 	print_version.c checksum.c vector.c linklist.c vty.c command.c \
 	sockunion.c prefix.c thread.c if.c memory.c buffer.c table.c hash.c \
@@ -155,9 +155,9 @@
 	debug.c sigevent.c
 
 
-libzebra_a_DEPENDENCIES = @LIB_REGEX@
+libzebra_so_DEPENDENCIES = @LIB_REGEX@
 
-libzebra_a_LIBADD = @LIB_REGEX@
+libzebra_so_LIBADD = @LIB_REGEX@
 
 pkginclude_HEADERS = \
 	buffer.h command.h filter.h getopt.h hash.h if.h linklist.h log.h \
@@ -175,8 +175,8 @@
 CONFIG_CLEAN_FILES = version.h
 LIBRARIES = $(lib_LIBRARIES)
 
-libzebra_a_AR = $(AR) cru
-am_libzebra_a_OBJECTS = version.$(OBJEXT) network.$(OBJEXT) \
+libzebra_so_AR = $(AR) cru
+am_libzebra_so_OBJECTS = version.$(OBJEXT) network.$(OBJEXT) \
 	pid_output.$(OBJEXT) getopt.$(OBJEXT) getopt1.$(OBJEXT) \
 	daemon.$(OBJEXT) print_version.$(OBJEXT) checksum.$(OBJEXT) \
 	vector.$(OBJEXT) linklist.$(OBJEXT) vty.$(OBJEXT) \
@@ -187,8 +187,8 @@
 	str.$(OBJEXT) log.$(OBJEXT) plist.$(OBJEXT) zclient.$(OBJEXT) \
 	sockopt.$(OBJEXT) smux.$(OBJEXT) md5.$(OBJEXT) \
 	if_rmap.$(OBJEXT) keychain.$(OBJEXT) privs.$(OBJEXT) \
-	debug.$(OBJEXT) sigevent.$(OBJEXT)
-libzebra_a_OBJECTS = $(am_libzebra_a_OBJECTS)
+	sigevent.$(OBJEXT)
+libzebra_so_OBJECTS = $(am_libzebra_so_OBJECTS)
 
 DEFAULT_INCLUDES =  -I. -I$(srcdir) -I$(top_builddir)
 depcomp = $(SHELL) $(top_srcdir)/depcomp
@@ -215,12 +215,12 @@
 	$(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS)
 CCLD = $(CC)
 LINK = $(CCLD) $(AM_CFLAGS) $(CFLAGS) $(AM_LDFLAGS) $(LDFLAGS) -o $@
-DIST_SOURCES = $(libzebra_a_SOURCES)
+DIST_SOURCES = $(libzebra_so_SOURCES)
 HEADERS = $(pkginclude_HEADERS)
 
 DIST_COMMON = $(pkginclude_HEADERS) $(srcdir)/Makefile.in ChangeLog \
 	Makefile.am version.h.in
-SOURCES = $(libzebra_a_SOURCES)
+SOURCES = $(libzebra_so_SOURCES)
 
 all: all-am
 
@@ -263,10 +263,9 @@
 
 clean-libLIBRARIES:
 	-test -z "$(lib_LIBRARIES)" || rm -f $(lib_LIBRARIES)
-libzebra.a: $(libzebra_a_OBJECTS) $(libzebra_a_DEPENDENCIES) 
-	-rm -f libzebra.a
-	$(libzebra_a_AR) libzebra.a $(libzebra_a_OBJECTS) $(libzebra_a_LIBADD)
-	$(RANLIB) libzebra.a
+libzebra.so: $(libzebra_so_OBJECTS) $(libzebra_so_DEPENDENCIES) 
+	-rm -f libzebra.so
+	$(LD) -o libzebra.so -shared $(libzebra_so_OBJECTS) $(libzebra_so_LIBADD)
 
 mostlyclean-compile:
 	-rm -f *.$(OBJEXT) core *.core
diff -ruNP quagga-0.96.5/lib/vty.c quagga-0.96.5-patched/lib/vty.c
--- quagga-0.96.5/lib/vty.c	Fri Apr 30 15:53:23 2004
+++ quagga-0.96.5-patched/lib/vty.c	Wed May  5 16:07:00 2004
@@ -2357,7 +2357,8 @@
 }
 
 /* Master of the threads. */
-static struct thread_master *master;
+//static struct thread_master *master;
+struct thread_master *master;
 
 static void
 vty_event (enum event event, int sock, struct vty *vty)
diff -ruNP quagga-0.96.5/zebra/Makefile.in quagga-0.96.5-patched/zebra/Makefile.in
--- quagga-0.96.5/zebra/Makefile.in	Mon May  3 15:03:09 2004
+++ quagga-0.96.5-patched/zebra/Makefile.in	Wed May  5 16:07:00 2004
@@ -170,7 +170,7 @@
 	interface.h ipforward.h
 
 
-zebra_LDADD = ../lib/libzebra.a $(otherobj) $(LIBCAP) $(LIB_IPV6)
+zebra_LDADD = -L../lib -lzebra $(otherobj) $(LIBCAP) $(LIB_IPV6)
 
 zebra_DEPENDENCIES = $(otherobj)
 
--- quagga-0.96.5/configure.orig	Thu Nov 18 11:25:18 2004
+++ quagga-0.96.5/configure	Thu Nov 18 11:46:44 2004
@@ -7546,7 +7546,7 @@
 if test "${enable_snmp}" = "yes";then
   if test "${HAVE_SNMP}" = ""; then
     old_libs="${LIBS}"
-    LIBS="-L/usr/lib"
+    LIBS=""
     unset ac_cv_lib_snmp_asn_parse_int
     echo "$as_me:$LINENO: checking for main in -lcrypto" >&5
 echo $ECHO_N "checking for main in -lcrypto... $ECHO_C" >&6
@@ -7723,7 +7723,7 @@
   fi
   if test "${HAVE_SNMP}" = ""; then
     old_libs="${LIBS}"
-    LIBS="-L/usr/lib"
+    LIBS=""
     unset ac_cv_lib_snmp_asn_parse_int
     echo "$as_me:$LINENO: checking for asn_parse_int in -lsnmp" >&5
 echo $ECHO_N "checking for asn_parse_int in -lsnmp... $ECHO_C" >&6
@@ -8082,13 +8082,13 @@
   fi
 
   if test "${HAVE_SNMP}" = "yes"; then
-    for ac_snmp in /usr/include/net-snmp/library/asn1.h /usr/include/ucd-snmp/asn1.h /usr/local/include/ucd-snmp/asn1.h /dev/null
+    for ac_snmp in ${UC_ROOT}/include/net-snmp/library/asn1.h /usr/include/net-snmp/library/asn1.h /usr/include/ucd-snmp/asn1.h /usr/local/include/ucd-snmp/asn1.h /dev/null
     do
       test -f "${ac_snmp}" && break
     done
 
     case ${ac_snmp} in
-      /usr/include/net-snmp/*)
+      */include/net-snmp/*)
 
 cat >>confdefs.h <<\_ACEOF
 #define HAVE_SNMP
@@ -8104,50 +8104,24 @@
 #define UCD_COMPATIBLE
 _ACEOF
 
-                  CFLAGS="${CFLAGS} -I/usr/include/net-snmp -I/usr/include/net-snmp/library"
+                  ac_snmp_inc_base=$(echo ${ac_snmp}|sed 's/^\(.*\/include\/net-snmp\).*$/\1/')
+                  CFLAGS="${CFLAGS} -I${ac_snmp_inc_base} -I${ac_snmp_inc_base}/library"
 		  if test "${HAVE_NETSNMP}" = "yes"; then
 		        LIBS="${LIBS} -lnetsnmp"
 		  else
 			LIBS="${LIBS} -lsnmp"
 		  fi
                   ;;
-      /usr/include/ucd-snmp/*)
+      */include/ucd-snmp/*)
 
 cat >>confdefs.h <<\_ACEOF
 #define HAVE_SNMP
 _ACEOF
 
-                  CFLAGS="${CFLAGS} -I/usr/include/ucd-snmp"
+                  ac_snmp_inc_base=$(echo ${ac_snmp}|sed 's/^\(.*\/include\/ucd-snmp\).*$/\1/')
+                  CFLAGS="${CFLAGS} -I${ac_snmp_inc_base}"
     		  LIBS="${LIBS} -lsnmp"
                   ;;
-      /usr/local/include/ucd-snmp/*)
-
-cat >>confdefs.h <<\_ACEOF
-#define HAVE_SNMP
-_ACEOF
-
-  	          CFLAGS="${CFLAGS} -I/usr/local/include/ucd-snmp"
-  		  LIBS="${LIBS} -L/usr/local/lib -lsnmp"
-                  ;;
-      /usr/local/include/net-snmp/*)
-
-cat >>confdefs.h <<\_ACEOF
-#define HAVE_SNMP
-_ACEOF
-
-
-cat >>confdefs.h <<\_ACEOF
-#define HAVE_NET_SNMP
-_ACEOF
-
-
-cat >>confdefs.h <<\_ACEOF
-#define UCD_COMPATIBLE
-_ACEOF
-
-      		CFLAGS="${CFLAGS} -I/usr/local/include/net-snmp"
-      		LIBS="${LIBS} -L/usr/local/lib -lnetsnmp"
-      		;;
     esac
     if test "${NEED_CRYPTO}" = "yes"; then
       LIBS="${LIBS} -lcrypto"
