--- quagga-0.96.5/configure.orig	Thu Nov 18 11:25:18 2004
+++ quagga-0.96.5/configure	Thu Nov 18 11:46:44 2004
@@ -7546,7 +7546,7 @@
 if test "${enable_snmp}" = "yes";then
   if test "${HAVE_SNMP}" = ""; then
     old_libs="${LIBS}"
-    LIBS="-L/usr/lib"
+    LIBS=""
     unset ac_cv_lib_snmp_asn_parse_int
     echo "$as_me:$LINENO: checking for main in -lcrypto" >&5
 echo $ECHO_N "checking for main in -lcrypto... $ECHO_C" >&6
@@ -7723,7 +7723,7 @@
   fi
   if test "${HAVE_SNMP}" = ""; then
     old_libs="${LIBS}"
-    LIBS="-L/usr/lib"
+    LIBS=""
     unset ac_cv_lib_snmp_asn_parse_int
     echo "$as_me:$LINENO: checking for asn_parse_int in -lsnmp" >&5
 echo $ECHO_N "checking for asn_parse_int in -lsnmp... $ECHO_C" >&6
@@ -8082,13 +8082,13 @@
   fi
 
   if test "${HAVE_SNMP}" = "yes"; then
-    for ac_snmp in /usr/include/net-snmp/library/asn1.h /usr/include/ucd-snmp/asn1.h /usr/local/include/ucd-snmp/asn1.h /dev/null
+    for ac_snmp in ${UC_ROOT}/include/net-snmp/library/asn1.h /usr/include/net-snmp/library/asn1.h /usr/include/ucd-snmp/asn1.h /usr/local/include/ucd-snmp/asn1.h /dev/null
     do
       test -f "${ac_snmp}" && break
     done
 
     case ${ac_snmp} in
-      /usr/include/net-snmp/*)
+      */include/net-snmp/*)
 
 cat >>confdefs.h <<\_ACEOF
 #define HAVE_SNMP
@@ -8104,50 +8104,24 @@
 #define UCD_COMPATIBLE
 _ACEOF
 
-                  CFLAGS="${CFLAGS} -I/usr/include/net-snmp -I/usr/include/net-snmp/library"
+                  ac_snmp_inc_base=$(echo ${ac_snmp}|sed 's/^\(.*\/include\/net-snmp\).*$/\1/')
+                  CFLAGS="${CFLAGS} -I${ac_snmp_inc_base} -I${ac_snmp_inc_base}/library"
 		  if test "${HAVE_NETSNMP}" = "yes"; then
 		        LIBS="${LIBS} -lnetsnmp"
 		  else
 			LIBS="${LIBS} -lsnmp"
 		  fi
                   ;;
-      /usr/include/ucd-snmp/*)
+      */include/ucd-snmp/*)
 
 cat >>confdefs.h <<\_ACEOF
 #define HAVE_SNMP
 _ACEOF
 
-                  CFLAGS="${CFLAGS} -I/usr/include/ucd-snmp"
+                  ac_snmp_inc_base=$(echo ${ac_snmp}|sed 's/^\(.*\/include\/ucd-snmp\).*$/\1/')
+                  CFLAGS="${CFLAGS} -I${ac_snmp_inc_base}"
     		  LIBS="${LIBS} -lsnmp"
                   ;;
-      /usr/local/include/ucd-snmp/*)
-
-cat >>confdefs.h <<\_ACEOF
-#define HAVE_SNMP
-_ACEOF
-
-  	          CFLAGS="${CFLAGS} -I/usr/local/include/ucd-snmp"
-  		  LIBS="${LIBS} -L/usr/local/lib -lsnmp"
-                  ;;
-      /usr/local/include/net-snmp/*)
-
-cat >>confdefs.h <<\_ACEOF
-#define HAVE_SNMP
-_ACEOF
-
-
-cat >>confdefs.h <<\_ACEOF
-#define HAVE_NET_SNMP
-_ACEOF
-
-
-cat >>confdefs.h <<\_ACEOF
-#define UCD_COMPATIBLE
-_ACEOF
-
-      		CFLAGS="${CFLAGS} -I/usr/local/include/net-snmp"
-      		LIBS="${LIBS} -L/usr/local/lib -lnetsnmp"
-      		;;
     esac
     if test "${NEED_CRYPTO}" = "yes"; then
       LIBS="${LIBS} -lcrypto"
