--- rp-pppoe-3.5/src/relay.c.orig	Tue Jun 21 12:03:25 2005
+++ rp-pppoe-3.5/src/relay.c	Tue Jun 21 12:04:14 2005
@@ -986,7 +986,7 @@
 
     /* Can a client legally be behind this interface? */
     if (!iface->clientOK) {
-	syslog(LOG_ERR,
+/*	syslog(LOG_ERR,
 	       "PADI packet from %02x:%02x:%02x:%02x:%02x:%02x on interface %s not permitted",
 	       packet->ethHdr.h_source[0],
 	       packet->ethHdr.h_source[1],
@@ -994,7 +994,7 @@
 	       packet->ethHdr.h_source[3],
 	       packet->ethHdr.h_source[4],
 	       packet->ethHdr.h_source[5],
-	       iface->name);
+	       iface->name); */
 	return;
     }
 
