--- rp-pppoe-3.5/src/pppoe-server.c.orig	2004-12-30 19:23:23.000000000 +0100
+++ rp-pppoe-3.5/src/pppoe-server.c	2004-12-30 19:25:02.000000000 +0100
@@ -1706,7 +1706,13 @@
     argv[c++] = "pppd";
     argv[c++] = "plugin";
     argv[c++] = PLUGIN_PATH;
-    argv[c++] = session->ethif->name;
+    snprintf(buffer, SMALLBUF, "nic-%s",
+	     session->ethif->name);
+    argv[c++] = strdup(buffer);
+    if (!argv[c-1]) {
+	/* TODO: Send a PADT */
+	exit(EXIT_FAILURE);
+    }
     snprintf(buffer, SMALLBUF, "%d:%02x:%02x:%02x:%02x:%02x:%02x",
 	     ntohs(session->sess),
 	     session->eth[0], session->eth[1], session->eth[2],
