--- rp-pppoe-3.5/src/pppoe-server.c.orig	Fri Mar 26 15:13:56 2004
+++ rp-pppoe-3.5/src/pppoe-server.c	Fri Mar 26 15:43:56 2004
@@ -1632,8 +1632,8 @@
 	     ntohs(session->sess),
 	     session->eth[0], session->eth[1], session->eth[2],
 	     session->eth[3], session->eth[4], session->eth[5],
-	     session->serviceName,
-	     PppoeOptions);
+	     PppoeOptions,
+	     session->serviceName);
     argv[c++] = strdup(buffer);
     if (!argv[c-1]) {
 	/* TODO: Send a PADT */
--- rp-pppoe-3.5/src/pppoe.c.orig	Mon Jul  5 10:16:11 2004
+++ rp-pppoe-3.5/src/pppoe.c	Mon Jul  5 10:16:45 2004
@@ -546,9 +546,9 @@
 #endif
     }
 
-    /* Set signal handlers: send PADT on HUP; ignore TERM and INT */
+    /* Set signal handlers: send PADT on HUP and TERM, ignore INT */
     if (!conn.printACNames) {
-	signal(SIGTERM, SIG_IGN);
+	signal(SIGTERM, sigPADT);
 	signal(SIGINT, SIG_IGN);
 	signal(SIGHUP, sigPADT);
 
