--- tcpdump-3.9.4/configure.orig	2006-06-20 16:24:53.000000000 +0200
+++ tcpdump-3.9.4/configure	2006-06-20 16:26:03.000000000 +0200
@@ -11504,7 +11504,7 @@
 	Xprefix=`eval echo $prefix`
 fi
 
-for dir in $Xprefix /usr/${host_alias} /usr /usr/local /usr/local/ssl /usr/pkg; do
+for dir in ; do
 	#
 	# XXX - is there a better way to check if a given library is
 	# in a given directory than checking each of the possible
