--- tcpdump-3.8.3/configure.orig	2005-02-06 00:32:03.000000000 +0100
+++ tcpdump-3.8.3/configure	2005-02-06 00:32:53.000000000 +0100
@@ -8879,18 +8879,10 @@
 int
 main ()
 {
-
-int
-return_pcap_debug(void)
-{
 	extern int pcap_debug;
 
 	return pcap_debug;
 }
-
-  ;
-  return 0;
-}
 _ACEOF
 rm -f conftest.$ac_objext conftest$ac_exeext
 if { (eval echo "$as_me:$LINENO: \"$ac_link\"") >&5
--- tcpdump-3.8.3/configure.orig	2005-02-06 00:36:05.000000000 +0100
+++ tcpdump-3.8.3/configure	2005-02-06 00:38:37.000000000 +0100
@@ -8930,17 +8930,9 @@
 int
 main ()
 {
-
-	int
-	return_yydebug(void)
-	{
 		extern int yydebug;
 
 		return yydebug;
-	}
-
-  ;
-  return 0;
 }
 _ACEOF
 rm -f conftest.$ac_objext conftest$ac_exeext
