--- uc/libc/misc/syslog/syslog.c.orig	2005-01-12 08:59:21.000000000 +0100
+++ uc/libc/misc/syslog/syslog.c	2005-12-27 18:17:39.000000000 +0100
@@ -79,6 +79,7 @@
 #include <stdio.h>
 #include <ctype.h>
 #include <signal.h>
+#include <fcntl.h>
 
 
 #ifdef __UCLIBC_HAS_THREADS__
@@ -289,7 +290,7 @@
 		UNLOCK;
 		return;
 	    }
-	    /*			fcntl(LogFile, F_SETFD, 1); */
+	fcntl(LogFile, F_SETFD, FD_CLOEXEC);
 	}
     }
 
