--- uClibc-0.9.26/libc/misc/syslog/syslog.c.orig	2005-06-04 20:28:10.000000000 +0200
+++ uClibc-0.9.26/libc/misc/syslog/syslog.c	2005-06-04 20:30:43.000000000 +0200
@@ -79,6 +79,7 @@
 #include <stdio.h>
 #include <ctype.h>
 #include <signal.h>
+#include <fcntl.h>
 
 
 #ifdef __UCLIBC_HAS_THREADS__
@@ -285,7 +286,7 @@
 		UNLOCK;
 		return;
 	    }
-	    /*			fcntl(LogFile, F_SETFD, 1); */
+	fcntl(LogFile, F_SETFD, FD_CLOEXEC);
 	}
     }
 
