--- uClibc-0.9.28/libc/misc/syslog/syslog.c.orig	2005-08-18 00:49:44.000000000 +0200
+++ uClibc-0.9.28/libc/misc/syslog/syslog.c	2006-06-14 19:42:29.000000000 +0200
@@ -79,6 +79,7 @@
 #include <stdio.h>
 #include <ctype.h>
 #include <signal.h>
+#include <fcntl.h>
 
 
 #ifdef __UCLIBC_HAS_THREADS__
@@ -293,7 +294,7 @@
 		UNLOCK;
 		return;
 	    }
-	    /*			fcntl(LogFile, F_SETFD, 1); */
+	fcntl(LogFile, F_SETFD, FD_CLOEXEC);
 	}
     }
 
