--- vtun/configure.orig	Mon Aug  2 02:50:52 2004
+++ vtun/configure	Mon Aug  2 02:51:04 2004
@@ -2112,7 +2112,7 @@
     echo $ac_n "checking "for blowfish.h"""... $ac_c" 1>&6
 echo "configure:2114: checking "for blowfish.h"" >&5 
     ac_hdr_found=no
-    for p in $BLOWFISH_HDR_DIR /usr/include/ssl /usr/include/openssl /usr/include /usr/local/include /usr/local/ssl/include  /usr/include/crypto; do
+    for p in $SSL_HDR_DIR /usr/include/ssl /usr/include/openssl /usr/include /usr/local/include /usr/local/ssl/include  /usr/include/crypto; do
 	if test -n "$p"; then
 	  dir="$p"
 	else
